File: scripts/sync-to-cloudflare.py
#!/usr/bin/env python3
"""
Cloudflare 동기화 스크립트
로컬 인덱스를 Cloudflare로 동기화
"""

import argparse
import json
import time
import requests
from typing import List, Dict, Any

# 설정
CHROMA_URL = "http://localhost:8000"
COLLECTION_NAME = "search-index"
DEFAULT_API_URL = "https://search-engine-api.pages.dev"

class CloudflareSync:
    def __init__(self, api_url: str, api_key: str):
        self.api_url = api_url
        self.api_key = api_key
        self.headers = {
            "Content-Type": "application/json",
            "X-API-Key": api_key
        }
        
    def test_connection(self) -> bool:
        """API 연결 테스트"""
        try:
            response = requests.get(f"{self.api_url}/api/health")
            if response.status_code == 200:
                print(f"✅ API 연결 성공: {self.api_url}")
                return True
            else:
                print(f"❌ API 연결 실패: {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ API 연결 오류: {e}")
            return False
    
    def sync_urls(self, urls: List[str], batch_size: int = 10) -> Dict[str, Any]:
        """URL 목록을 Cloudflare로 동기화"""
        results = {"success": 0, "failed": 0, "total": len(urls)}
        
        for i in range(0, len(urls), batch_size):
            batch = urls[i:i+batch_size]
            print(f"\n[{i+1}-{min(i+batch_size, len(urls))}/{len(urls)}] 동기화 중...")
            
            try:
                response = requests.post(
                    f"{self.api_url}/api/index",
                    headers=self.headers,
                    json={"urls": batch}
                )
                
                if response.status_code == 200:
                    data = response.json()
                    if data.get("success"):
                        results["success"] += len(batch)
                        print(f"✅ 배치 동기화 성공: {len(batch)}개 URL")
                    else:
                        results["failed"] += len(batch)
                        print(f"❌ 배치 동기화 실패: {data.get('error', 'Unknown error')}")
                else:
                    results["failed"] += len(batch)
                    print(f"❌ 배치 동기화 실패: HTTP {response.status_code}")
                
            except Exception as e:
                results["failed"] += len(batch)
                print(f"❌ 배치 동기화 오류: {e}")
            
            # Rate limit 방지
            time.sleep(2)
        
        return results

def main():
    parser = argparse.ArgumentParser(description="Cloudflare 동기화 스크립트")
    parser.add_argument("--api-url", default=DEFAULT_API_URL,
                       help="Cloudflare API URL")
    parser.add_argument("--api-key", required=True,
                       help="API 키")
    parser.add_argument("--urls", nargs="+",
                       help="동기화할 URL 목록")
    parser.add_argument("--batch-size", type=int, default=10,
                       help="배치 크기")
    
    args = parser.parse_args()
    
    # 동기화 초기화
    sync = CloudflareSync(args.api_url, args.api_key)
    
    # 연결 테스트
    if not sync.test_connection():
        print("❌ API 연결 실패. URL과 API 키를 확인하세요.")
        return
    
    # URL 목록 준비
    if not args.urls:
        print("❌ URL 목록을 지정하세요. (--urls 옵션)")
        return
    
    # 동기화 실행
    print(f"\n🚀 Cloudflare 동기화 시작")
    print(f"   API URL: {args.api_url}")
    print(f"   URL 수: {len(args.urls)}")
    print()
    
    start_time = time.time()
    results = sync.sync_urls(args.urls, args.batch_size)
    elapsed = time.time() - start_time
    
    # 결과 출력
    print(f"\n{'='*60}")
    print(f"✅ 동기화 완료!")
    print(f"   성공: {results['success']}/{results['total']}")
    print(f"   실패: {results['failed']}/{results['total']}")
    print(f"   소요 시간: {elapsed:.1f}초")
    print(f"{'='*60}")

if __name__ == "__main__":
    main()

File: scripts/sync-to-cloudflare.ts
#!/usr/bin/env -S npx tsx
/**
 * 로컬 → Cloudflare 동기화 스크립트
 * 
 * 로컬 ChromaDB의 인덱스를 Cloudflare Vectorize + D1로 동기화
 * 
 * 사용법:
 *   npx tsx scripts/sync-to-cloudflare.ts --api-url=https://search-engine-api.pages.dev
 *   npx tsx scripts/sync-to-cloudflare.ts --api-url=https://search-engine-api.pages.dev --dry-run
 */

import { LocalIndexingService } from '../src/lib/local-indexing'

// ============================================================
// Configuration
// ============================================================

interface SyncArgs {
  apiUrl: string
  apiKey?: string
  chromaUrl: string
  ollamaUrl: string
  batchSize: number
  dryRun: boolean
}

function parseArgs(): SyncArgs {
  const args = process.argv.slice(2)
  const options: SyncArgs = {
    apiUrl: 'https://search-engine-api.pages.dev',
    chromaUrl: 'http://localhost:8000',
    ollamaUrl: 'http://localhost:11434',
    batchSize: 20,
    dryRun: false,
  }

  for (const arg of args) {
    if (arg.startsWith('--api-url=')) {
      options.apiUrl = arg.split('=')[1]
    } else if (arg.startsWith('--api-key=')) {
      options.apiKey = arg.split('=')[1]
    } else if (arg.startsWith('--chroma=')) {
      options.chromaUrl = arg.split('=')[1]
    } else if (arg.startsWith('--ollama=')) {
      options.ollamaUrl = arg.split('=')[1]
    } else if (arg.startsWith('--batch-size=')) {
      options.batchSize = parseInt(arg.split('=')[1], 10)
    } else if (arg === '--dry-run') {
      options.dryRun = true
    }
  }

  return options
}

// ============================================================
// Main Sync Logic
// ============================================================

async function syncToCloudflare(args: SyncArgs) {
  console.log('═══════════════════════════════════════════════════════')
  console.log('  로컬 → Cloudflare 동기화')
  console.log('═══════════════════════════════════════════════════════')
  console.log(`  API URL: ${args.apiUrl}`)
  console.log(`  ChromaDB: ${args.chromaUrl}`)
  console.log(`  배치 크기: ${args.batchSize}`)
  console.log(`  Dry run: ${args.dryRun}`)
  console.log('')

  // 로컬 서비스 초기화
  const localService = new LocalIndexingService({
    chromaUrl: args.chromaUrl,
    ollamaUrl: args.ollamaUrl,
  })

  const status = await localService.initialize()
  if (!status.chroma) {
    console.error('❌ ChromaDB 미연결')
    process.exit(1)
  }

  // 로컬 인덱스 상태 확인
  const localStats = await localService.getStats()
  console.log(`  로컬 청크 수: ${localStats.totalChunks}`)
  console.log('')

  if (localStats.totalChunks === 0) {
    console.log('로컬 인덱스가 비어있습니다. 먼저 인덱싱을 실행하세요.')
    console.log('  npx tsx scripts/local-index.ts --category=tech')
    return
  }

  // Cloudflare에서 기존 인덱스 상태 확인
  console.log('Cloudflare 인덱스 상태 확인 중...')
  try {
    const response = await fetch(`${args.apiUrl}/api/index/stats`)
    const stats = await response.json() as { totalDocuments?: number }
    console.log(`  Cloudflare 문서 수: ${stats.totalDocuments ?? 0}`)
  } catch (err) {
    console.warn('⚠️ Cloudflare 상태 확인 실패:', err)
  }
  console.log('')

  // 동기화 실행
  if (args.dryRun) {
    console.log('[Dry Run] 동기화할 URL 목록:')
    console.log('  (실제 동기화를 위해 --dry-run 옵션을 제거하세요)')
    return
  }

  console.log('동기화 시작...')

  // ChromaDB에서 모든 문서 가져오기
  // 주의: ChromaDB API는 전체 문서 가져오기를 제한할 수 있음
  // 실제 구현에서는 페이지네이션이 필요할 수 있음

  // 임시: 샘플 URL 목록으로 테스트
  const sampleUrls = [
    'https://react.dev/learn',
    'https://react.dev/reference/react',
    'https://vuejs.org/guide/introduction.html',
    'https://nextjs.org/docs',
    'https://www.typescriptlang.org/docs/',
  ]

  console.log(`동기화할 URL: ${sampleUrls.length}개`)

  // 배치로 Cloudflare API 전송
  let successCount = 0
  let failCount = 0

  for (let i = 0; i < sampleUrls.length; i += args.batchSize) {
    const batch = sampleUrls.slice(i, i + args.batchSize)
    const progress = `[${Math.floor(i / args.batchSize) + 1}/${Math.ceil(sampleUrls.length / args.batchSize)}]`

    process.stdout.write(`${progress} 배치 동기화 중... `)

    try {
      const headers: Record<string, string> = {
        'Content-Type': 'application/json',
      }
      if (args.apiKey) {
        headers['X-API-Key'] = args.apiKey
      }

      const response = await fetch(`${args.apiUrl}/api/index`, {
        method: 'POST',
        headers,
        body: JSON.stringify({ urls: batch }),
      })

      if (response.ok) {
        const result = await response.json() as { stats?: { succeeded: number; failed: number } }
        successCount += result.stats?.succeeded ?? 0
        failCount += result.stats?.failed ?? 0
        console.log(`✅ ${result.stats?.succeeded ?? 0} 성공`)
      } else {
        console.log(`❌ HTTP ${response.status}`)
        failCount += batch.length
      }
    } catch (_err) {
      console.log(`❌ 에러`)
      failCount += batch.length
    }
  }

  console.log('')
  console.log('═══════════════════════════════════════════════════════')
  console.log(`  ✅ 완료: ${successCount} 성공, ${failCount} 실패`)
  console.log('═══════════════════════════════════════════════════════')
}

// ============================================================
// Run
// ============================================================

const args = parseArgs()
syncToCloudflare(args).catch(console.error)
